### PR TITLE
[USM] Support index 4 in http2 decoding

### PR DESCRIPTION
**Pull request by amitslavin**
### What does this PR do?

Supports the use case where the path is sent with index 4.

### Motivation

Supports all possible cases in gRPC/HTTP2 decoding to improve accuracy.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/usm/monitor_test.go [hunk 1]
```diff
@@ -1001,6 +1014,7 @@ func (s *USMHTTP2Suite) TestRawTraffic() {
 		name                  string
 		numberOfSettingFrames int
 		numberOfRequestFrames int
+		isEmptyPath           bool
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:1017]: let's rename it
or let's get the buffer as a parameters

File: pkg/network/usm/monitor_test.go [hunk 2]
```diff
@@ -972,12 +972,25 @@ var (
 		0xab, 0x7a, 0x8d, 0xc4, 0x75, 0xa7, 0x4a, 0x6b, 0x58, 0x94, 0x18, 0xb5, 0x25, 0x81, 0x2e, 0x0f,
 	}
 
+	emptyRequest = []byte{
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:975]: let's rename this buffer

File: pkg/network/usm/monitor_test.go [hunk 3]
```diff
@@ -972,12 +972,25 @@ var (
...
 		0x00, 0x00, 0x04, 0x00, 0x01, 0x00, 0x00, 0x00, 0x03, 0x74, 0x65, 0x73, 0x74,
 	}
 )
 
+// createRawEmptyRequestFrame creates a raw request frame with the given streamID and path index 4 ("/").
+func createRawEmptyRequestFrame(streamID int) []byte {
+	binary.BigEndian.PutUint16(emptyRequest[7:9], uint16(streamID))
+	return emptyRequest
+}
+
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:993]: as `createRawDataFrame`, `createRawRequestFrame`, `createRawEmptyRequestFrame` are all the same, and only the buffer is the different part - you can generalize the method as:

```go
func changeStreamIDInFrame(buf []byte, streamID int) []byte {
	binary.BigEndian.PutUint16(buf[7:9], uint16(streamID))
	return buf
}
```

File: pkg/network/usm/monitor_test.go [hunk 4]
```diff
@@ -1063,6 +1077,19 @@ func (s *USMHTTP2Suite) TestRawTraffic() {
...
 		},
+		{
+			name: "validate path with index 4",
+			// The purpose of this test is to verify our ability to identify paths with index 4.
+			numberOfSettingFrames: 0,
+			numberOfRequestFrames: 5,
+			isEmptyPath:           true,
+			expectedEndpoints: map[http.Key]int{
+				{
+					Path:   http.Path{Content: http.Interner.GetString("/")},
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:1088]: this is a fully indexed path and not literal one

**Review by guyarb — CHANGES_REQUESTED** [review]

**Review by guyarb — APPROVED** [review]
LGTM

File: pkg/network/usm/monitor_test.go [hunk 5]
```diff
@@ -991,6 +991,54 @@ func generateTestHeaderFields(pathNeverIndexed, withoutIndexing bool) []hpack.He
...
+	headerFields = removeHeaderFieldByKey(headerFields, ":path")
+
+	for i := 0; i < headersCount; i++ {
+		streamID := 2*i + 1
+		headersFrame, err := usmhttp2.NewHeadersFrameMessage(headerFields)
+		require.NoError(t, err, "could not create headers frame")
+
+		// we are adding the path header field with index 4, we need to do it on the byte slice and not on the headerFields
+		// due to the fact that when we create a header field it would be with index 5.
+		headersFrame = append(pathHeaderField, headersFrame...)
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:1025]: You're adding a path header on every header

File: pkg/network/usm/monitor_test.go [hunk 6]
```diff
@@ -991,6 +991,54 @@ func generateTestHeaderFields(pathNeverIndexed, withoutIndexing bool) []hpack.He
...
+	// pathHeaderField is the hex representation of the path /aaa with index 4.
+	pathHeaderField := []byte{0x44, 0x83, 0x60, 0x63, 0x1f}
+
+	// we remove the header field with key ":path" so that we will not have two header fields.
+	headerFields = removeHeaderFieldByKey(headerFields, ":path")
+
+	for i := 0; i < headersCount; i++ {
+		streamID := 2*i + 1
+		headersFrame, err := usmhttp2.NewHeadersFrameMessage(headerFields)
+		require.NoError(t, err, "could not create headers frame")
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:1021]: it does not change for all iterations, so you can (and should) move it to be calculated once

File: pkg/network/usm/monitor_test.go [hunk 7]
```diff
@@ -991,6 +991,54 @@ func generateTestHeaderFields(pathNeverIndexed, withoutIndexing bool) []hpack.He
...
+			break
+		}
+	}
+
+	return headerFields
+}
+
+// createHeadersWithIndexedPathKey creates a message with the given number of header frames but the
+// path frame header key 4 and not 5 as usual.
+func createHeadersWithIndexedPathKey(t *testing.T, headerFields []hpack.HeaderField, headersCount int) []byte {
```
Comment (guyarb) [inline pkg/network/usm/monitor_test.go:1009]:
```suggestion
func createHeadersWithIndexedPathKey(t *testing.T, headerFields []hpack.HeaderField, headerFramesCount int) []byte {
```

File: pkg/network/usm/monitor_test.go (hunk 5 above)
Comment (guyarb) [reply pkg/network/usm/monitor_test.go:1025]: similar to https://github.com/DataDog/datadog-agent/pull/21909/files#r1453371149

Comment (amitslavin) [reply pkg/network/usm/monitor_test.go:1025]: The change should solve the confusion over here

**Review by guyarb — APPROVED** [review]

**Pull request merged**